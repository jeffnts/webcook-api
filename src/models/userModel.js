import mongoose from 'mongoose'
const Schema = mongoose.Schema
import bcrypt from 'bcryptjs'

const userSchema = new Schema({

    name:{
        type: String,
        required: true
    },
    userName:{
        type: String,
        required: true,
        unique: true,
        lowercase: true,
        trim: true
    },
    email : {
        type: String,
        unique: true,
        required: true,
        lowercase: true,
        validate: function(email) {
            return /^[a-zA-Z0-9.!#$%&â€™*+\/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(email)
        }

    },
    password:{
        type: String,
        required: true,
        select: false
    },
    role:{
        type: String,
        required: true,
        enum:[
            'user',
            'admin'
        ],
        default: 'user'
    },
    photo:{
        type: String,
        default: 'avatar'
    },
    recipes: [
        {   type: Schema.Types.ObjectId,
            ref: 'Recipe'

        }

    ]


});
// Auth settings with Brcypt
    userSchema.pre('save', function hashPassword(next) {
        if (this.isModified('password')) {
            this.encryptPassword(this.password)
                .then(hash => {
                    this.password = hash;
                    next();
                })
                .catch(err => next(err));
        } else {
            return next();
        }
    });

    userSchema.methods = {
        async authenticate(plainText) {
            try {
                return await bcrypt.compare(plainText, this.password);
            } catch (err) {
                return false;
            }
        },
        encryptPassword(password) {
            return bcrypt.hash(password, 8);
        },
    };


export default  mongoose.model('User', userSchema);