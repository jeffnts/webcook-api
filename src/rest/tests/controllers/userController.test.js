//import app from '../../../app'
const app = require('../../../app')
const request = require('supertest')(app.callback())
import userModel from '../../../models/userModel'
const connection = require('../../../config/database/database_tests')
const {sanitizeTestObject} = require('../../../config/tests/sanitizeTestObject')

const dropUserCollection = async () => userModel.remove()


let user = {
    '_id': 'TESTOBJECT01',
    'name': 'Jhon',
    'email': 'jhon@email.com',
    'userName': 'jhon',
    'password': '123456'
}

const userLogin = {
  'userName': 'jhon',
  'password': '123456'
};

beforeAll(() => {

dropUserCollection()

});

afterAll(() => {

connection.disconect()
});



describe('Users Tests', () => {

    // describe('GET Method with no users registered.', () => {
    //
    //     test('It should return no one user.', async () => {
    //
    //         const res = await request.get('/api/users');
    //         expect(sanitizeTestObject(res.body)).toMatchSnapshot();
    //         expect(res.status).toEqual(403);
    //
    //
    //     });
    // });

    describe('POST method', () => {
        test('It should create a new user.', async () => {
           const res = await request.post('/api/user').send(user)
           expect(sanitizeTestObject(res.body)).toMatchSnapshot()
           expect(res.status).toEqual(201)


        });
        test('It should not create a new user with the same email.', async () => {

          const res = await request.post('/api/user').send(user)
          expect(sanitizeTestObject(res.body)).toMatchSnapshot()
          expect(res.status).toEqual(403)

        });
        test('It should not create a new user with  a already exists username.', async () => {

          user.email = "other@email.com"
          const res = await request.post('/api/user').send(user)
          expect(sanitizeTestObject(res.body)).toMatchSnapshot()
          expect(res.status).toEqual(406)

        });
    });

    describe('GET method', () => {

        test('It should not return all users if the user is not a admin.', async () => {

          let tokenRequest = await request.post('/api/auth/login').send(userLogin)
          let token = tokenRequest.body.token
          const res = await request.get('/api/users').set('token', token)
          expect(sanitizeTestObject(res.body)).toMatchSnapshot()
          expect(res.status).toEqual(403)

        });

        test('It should not return a user if he is not logged.', async () => {

          const res = await request.get('/api/user');
          expect(sanitizeTestObject(res.body)).toMatchSnapshot();
          expect(res.status).toEqual(401);

        });
        test('It should return a logged user according the tokken passed as parameter in the Header.', async () => {

          let tokenRequest = await request.post('/api/auth/login').send(userLogin)
          let token = tokenRequest.body.token
          const res = await request.get('/api/user').set('token', token)
          expect(sanitizeTestObject(res.body)).toMatchSnapshot()
          expect(res.status).toEqual(200)

        });
    });

    describe('PUT method', () => {

        test('It should not update a user if he is not logged.', async () => {

          const res = await request.put('/api/user').send({'name': 'Jhon Doe'})
          expect(sanitizeTestObject(res.body)).toMatchSnapshot()
          expect(res.status).toEqual(401)
        });
        test('It should update a logged user according the token passed as parameter in the Header.', async () => {

          let tokenRequest = await request.post('/api/auth/login').send(userLogin)
          let token = tokenRequest.body.token
          const res = await request.put('/api/user').send({'name': 'Jhon Doe'}).set('token', token)
          expect(sanitizeTestObject(res.body)).toMatchSnapshot()
          expect(res.status).toEqual(200)

        });
    });

    describe('DELETE method', () => {
        test('It should not delete a user if is not logged.', async () => {

          const res = await request.delete('/api/user')
          expect(sanitizeTestObject(res.body)).toMatchSnapshot()
          expect(res.status).toEqual(401)

        });
        test('It should delete a logged user according the token passed as parameter in the Header.', async () => {

          let tokenRequest = await request.post('/api/auth/login').send(userLogin)
          let token = tokenRequest.body.token
          const res = await request.delete('/api/user').set('token', token)
          expect(sanitizeTestObject(res.body)).toMatchSnapshot()
          expect(res.status).toEqual(204)

        });

    });
});