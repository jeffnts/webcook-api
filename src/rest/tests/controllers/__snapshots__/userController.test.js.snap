// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Tests DELETE method It should delete a logged user according the token passed as parameter in the Header. 1`] = `Object {}`;

exports[`Users Tests DELETE method It should not delete a user if is not logged. 1`] = `Object {}`;

exports[`Users Tests GET method It should not return a user if he is not logged. 1`] = `Object {}`;

exports[`Users Tests GET method It should not return all users if the user is not a admin. 1`] = `Object {}`;

exports[`Users Tests GET method It should return a logged user according the tokken passed as parameter in the Header. 1`] = `
Object {
  "__v": 0,
  "_id": "ObjectId",
  "email": "jhon@email.com",
  "name": "Jhon",
  "photo": "avatar",
  "recipes": Array [],
  "role": "user",
  "userName": "jhon",
}
`;

exports[`Users Tests POST method It should create a new user. 1`] = `Object {}`;

exports[`Users Tests POST method It should not create a new user with  a already exists username. 1`] = `Object {}`;

exports[`Users Tests POST method It should not create a new user with the same email. 1`] = `Object {}`;

exports[`Users Tests PUT method It should not update a user if he is not logged. 1`] = `Object {}`;

exports[`Users Tests PUT method It should update a logged user according the token passed as parameter in the Header. 1`] = `Object {}`;
